runtime/oti/FFIUpcallThunkGenHelpers.hpp

```diff
@@ -0,0 +1,300 @@
+ * Format: "IIHF R1,I2"
+ * R1 - destinationRegister
+ * I2 - integerConstant
+ *
+ * Returns: instruction size in bytes
+ */
+static I_32
+IIHF(I_8 *instructionPtr, I_32 destinationRegister, I_32 integerConstant)
+{
+    I_32 opcode = 0xC008;
+    *((I_16 *)instructionPtr) = (opcode | (I_16)(destinationRegister << 4));
+    instructionPtr += 2;
+    *((I_32 *)instructionPtr) = integerConstant;
+    return 6;
+}
+
+/**
+ * Generates an instruction to materialize a 32-bit integer constant into the lower-half R1
+ * Format: "IILF R1, I2"
+ * R1 - destinationRegister
+ * I2 - integerConstant
+ *
+ * Returns: instruction size in bytes
+ */
+static I_32
+IILF(I_8 *instructionPtr, I_32 destinationRegister, I_32 integerConstant)
+{
+    I_32 opcode = 0xC009;
+    *((I_16 *)instructionPtr) = (opcode | (I_16)(destinationRegister << 4));
+    instructionPtr += 2;
+    *((I_32 *)instructionPtr) = integerConstant;
+    return 6;
+}
+
+/**
+ * Generates an instruction to load a 64-bit value into R1, from the memory destination specified by X2, B2, and D2.
+ * Format: "LG R1,D2(X2,B2)"
+ * R1 - destinationRegister
+ * D2 - displacementInteger
+ * X2 - offsetRegister
+ * B2 - baseRegister
+ *
+ * Returns: instruction size in bytes
+ */
+static I_32
+LG(I_8 *instructionPtr, I_32 destinationRegister, I_32 offsetRegister, I_32 baseRegister, I_32 displacementInteger)
+{
+    I_32 opcode = 0xE3000000;
+    *((I_32 *)instructionPtr) = (opcode | (destinationRegister << 20) | (offsetRegister << 16) | (baseRegister << 12) | (displacementInteger & 0x00000FFF));
+    instructionPtr[4] = (I_8)((displacementInteger & 0x000FF000) >> 12);
+    instructionPtr[5] = 0x04;
+    return 6;
+}
+
+/**
+ * Generates an instruction to store a 64-bit value from R1, into the memory destination specified by X2, B2, and D2.
+ * Format: "STG R1,D2(X2,B2)"
+ * R1 - sourceRegister
+ * D2 - displacementInteger
+ * X2 - offsetRegister
+ * B2 - baseRegister
+ *
+ * Returns: instruction size in bytes
+ */
+static I_32
+STG(I_8 *instructionPtr, I_32 sourceRegister, I_32 offsetRegister, I_32 baseRegister, I_32 displacementInteger)
+{
+    I_32 opcode = 0xE3000000;
+    *((I_32 *)instructionPtr) = (opcode | (sourceRegister << 20) | (offsetRegister << 16) | (baseRegister << 12) | (displacementInteger & 0x00000FFF));
+    instructionPtr[4] = (I_8)((displacementInteger & 0x000FF000) >> 12);
+    instructionPtr[5] = 0x24;
+    return 6;
+}
+
+/*
+ * Generates an instruction that moves 'L' bytes of data from the second operand location(specified by D2 and B2) to the first operand location (specified by D1,B1).
+ * Format: D1(L,B1),D2(B2)
+ * D1 - destinationDisplacementInt
+ * B1 - destinationRegister
+ * D2 - sourceDisplacementInt
+ * B2 - sourceRegister
+ * L -  length
+ */
+static I_32
+MVC(I_8 *instructionPtr, I_32 destinationDisplacementInt, I_32 length, I_32 destinationRegister, I_32 sourceDisplacementInt, I_32 sourceRegister)
+{
+	I_32 instructionSize = 6;
```

Comment:
I understand carrying this from the previous implementation, can we also fix the return (instructionSize) here like what we did for previous instructions ?